src/items.rs

```diff
@@ -643,8 +645,27 @@ impl<'a> FmtVisitor<'a> {
         let mut items: Vec<_> = itemize_list_with(self.config.struct_variant_width());
 
         // If one of the variants use multiple lines, use multi-lined formatting for all variants.
-        let has_multiline_variant = items.iter().any(|item| item.inner_as_ref().contains('\n'));
-        let has_single_line_variant = items.iter().any(|item| !item.inner_as_ref().contains('\n'));
+        fn is_multi_line_variant(item: &ListItem) -> bool {
+            let variant_str = item.inner_as_ref();
+            let mut first_line_is_read = false;
+            for line in variant_str.split('\n') {
+                if first_line_is_read {
+                    return false;
+                }
+
+                // skip rustdoc comments and macro attributes
+                let line = line.trim_start();
+                if line.starts_with("///") || line.starts_with("#") {
+                    continue;
+                } else {
+                    first_line_is_read = true;
+                }
+            }
+
+            true
+        }
+        let has_multiline_variant = items.iter().any(is_multi_line_variant);
+        let has_single_line_variant = items.iter().any(|item| !is_multi_line_variant(item));
```

Comment:
[Any change we make would need to be gated](https://github.com/rust-lang/rustfmt/blob/master/Contributing.md#gate-formatting-changes). We can't change existing formatting because of rustfmt's stability guarantees.

Reply:
Thank you. I gated the changes in 103d13e0.